src/TurbulenceClosures/turbulence_closure_implementations/Smagorinskys/dynamic_coefficient.jl

```diff
@@ -0,0 +1,248 @@
+
+    return LM_ijk, MM_ijk
+end
+
+#####
+##### Directionally-averaged functionality
+#####
+
+function compute_coefficient_fields!(diffusivity_fields, closure::DirectionallyAveragedDynamicSmagorinsky, model; parameters)
+    grid = model.grid
+    arch = architecture(grid)
+    velocities = model.velocities
+    cˢ = closure.coefficient
+
+    if cˢ.schedule(model)
+        Σ = diffusivity_fields.Σ
+        Σ̄ = diffusivity_fields.Σ̄
+        launch!(arch, grid, :xyz, _compute_Σ_Σ̄!, Σ, Σ̄, grid, velocities...)
+
+        LM = diffusivity_fields.LM
+        MM = diffusivity_fields.MM
+        launch!(arch, grid, :xyz, _compute_LM_MM!, LM, MM, Σ, Σ̄, grid, velocities...)
+
+        𝒥ᴸᴹ = diffusivity_fields.𝒥ᴸᴹ
+        𝒥ᴹᴹ = diffusivity_fields.𝒥ᴹᴹ
+        compute!(𝒥ᴸᴹ)
+        compute!(𝒥ᴹᴹ)
+    end
+
+    return nothing
+end
+
+function allocate_coefficient_fields(closure::DirectionallyAveragedDynamicSmagorinsky, grid)
+    LM = CenterField(grid)
+    MM = CenterField(grid)
+
+    Σ = CenterField(grid)
+    Σ̄ = CenterField(grid)
+
+    𝒥ᴸᴹ = Field(Average(LM, dims=closure.coefficient.averaging))
+    𝒥ᴹᴹ = Field(Average(MM, dims=closure.coefficient.averaging))
+
+    return (; Σ, Σ̄, LM, MM, 𝒥ᴸᴹ, 𝒥ᴹᴹ)
+end
+
+#####
+##### Lagrangian-averaged functionality
+#####
+
+const c = Center()
+
+@kernel function _lagrangian_average_LM_MM!(𝒥ᴸᴹ, 𝒥ᴹᴹ, 𝒥ᴸᴹ⁻, 𝒥ᴹᴹ⁻, 𝒥ᴸᴹ_min, Σ, Σ̄, grid, Δt, u, v, w)
+    i, j, k = @index(Global, NTuple)
+    LMⁿ, MMⁿ = LM_and_MM(i, j, k, grid, Σ, Σ̄, u, v, w)
+    FT = eltype(grid)
+
+    @inbounds begin
+        𝒥ᴸᴹ⁻ᵢ = max(𝒥ᴸᴹ⁻[i, j, k], 𝒥ᴸᴹ_min)
+        𝒥ᴹᴹ⁻ᵢ = 𝒥ᴹᴹ⁻[i, j, k]
+
+        # Compute time scale
+        𝒥ᴸᴹ𝒥ᴹᴹ = 𝒥ᴸᴹ⁻ᵢ * 𝒥ᴹᴹ⁻ᵢ
+
+        if 𝒥ᴸᴹ𝒥ᴹᴹ < 0
+            @show 𝒥ᴸᴹ⁻ᵢ 𝒥ᴹᴹ⁻ᵢ 
+        end
+
+        eighth_root_𝒥ᴸᴹ𝒥ᴹᴹ = sqrt(sqrt(sqrt(sqrt(𝒥ᴸᴹ𝒥ᴹᴹ))))
+        T⁻ = convert(FT, 1.5) * Δᶠ(i, j, k, grid) / eighth_root_𝒥ᴸᴹ𝒥ᴹᴹ
+        τ = Δt / T⁻
+        ϵ = τ / (1 + τ)
+                        
+        # Compute interpolation
+        x = xnode(i, j, k, grid, c, c, c)
+        y = ynode(i, j, k, grid, c, c, c)
+        z = znode(i, j, k, grid, c, c, c)
+
+        # Displacements
+        δx = u[i, j, k] * Δt
+        δy = v[i, j, k] * Δt
+        δz = w[i, j, k] * Δt
+
+        # Prevent displacements from getting too big?
+        Δx = Δxᶜᶜᶜ(i, j, k, grid)
+        Δy = Δyᶜᶜᶜ(i, j, k, grid)
+        Δz = Δzᶜᶜᶜ(i, j, k, grid)
+
+        δx = clamp(δx, -Δx, Δx)
+        δy = clamp(δy, -Δy, Δy)
+        δz = clamp(δz, -Δz, Δz)
+
+        # Previous locations
+        x⁻ = x - δx
+        y⁻ = y - δy
+        z⁻ = z - δz
+        X⁻ = (x⁻, y⁻, z⁻)
+
+        itp_𝒥ᴹᴹ = interpolate(X⁻, 𝒥ᴹᴹ⁻, (c, c, c), grid)
+        itp_𝒥ᴸᴹ = interpolate(X⁻, 𝒥ᴸᴹ⁻, (c, c, c), grid)
+
+        # Take time-step
+        𝒥ᴹᴹ[i, j, k] = ϵ * MMⁿ + (1 - ϵ) * itp_𝒥ᴹᴹ
+
+        𝒥ᴸᴹ★ = ϵ * LMⁿ + (1 - ϵ) * max(itp_𝒥ᴸᴹ, 𝒥ᴸᴹ_min)
+        𝒥ᴸᴹ[i, j, k] = max(𝒥ᴸᴹ★, 𝒥ᴸᴹ_min)
```

Comment:
@glwagner is the first `max()` needed?

Reply:
I believe so yes

Reply:
The first is the important one. The second matters at initialization.